slidechain/mockhorizon/mockhorizon.go

```diff
@@ -0,0 +1,146 @@
+package mockhorizon
+
+import (
+	"context"
+	"sync"
+
+	"github.com/pkg/errors"
+	"github.com/stellar/go/clients/horizon"
+	"github.com/stellar/go/network"
+	"github.com/stellar/go/xdr"
+)
+
+// New returns a new instance of the mockhorizon
+// struct, which implements horizon.ClientInterface
+func New() *Client {
+	return &Client{
+		mu:        new(sync.Mutex),
+		submitted: sync.NewCond(new(sync.Mutex)),
+	}
+}
+
+type Client struct {
+	txs       []string
+	mu        *sync.Mutex
+	submitted *sync.Cond
+}
+
+// Implement horizon.ClientInterface for mockhorizon.
+//
+// Our mock implementation assumes that the calling functions
+// want to submit transactions and then stream to see if they
+// have been successfully included in the ledger.
+
+// SubmitTransaction unmarshals the tx envelope string into a xdr.TransactionEnvelope,
+// and then adds the transaction to the Client's internal record of transactions to
+// "stream".
+func (c *Client) SubmitTransaction(txeBase64 string) (horizon.TransactionSuccess, error) {
+	var txe xdr.TransactionEnvelope
+	err := xdr.SafeUnmarshalBase64(txeBase64, &txe)
+	if err != nil {
+		return horizon.TransactionSuccess{}, errors.Wrap(err, "submittx: unmarshaling tx envelope")
+	}
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.txs = append(c.txs, txeBase64)
+	c.submitted.Broadcast()
+	return horizon.TransactionSuccess{}, nil
+}
+
+// StreamTransactions "streams" the transactions that have been submitted to SubmitTransaction.
+func (c *Client) StreamTransactions(ctx context.Context, accountID string, cursor *horizon.Cursor, handler horizon.TransactionHandler) error {
+	txindex := 0
+	ch := make(chan struct{})
+
+	go func() {
+		c.submitted.L.Lock()
+		defer c.submitted.L.Unlock()
+		for {
+			if ctx.Err() != nil {
+				return
+			}
+			c.submitted.Wait()
+			ch <- struct{}{}
+		}
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-ch:
+		}
+
+		// figure out how we exactly want to "stream" txs
+		// stream all the ones in the array
+		// and then wait for new ones?
```

Comment:
This approach does make sense to me. My only concern is that this could end up storing more things in memory than we'd like, and something like a queue could avoid that. I don't think that matters in the context of unit tests.

Reply:
oh man sorry for leaving this comment to myself in there! I was thinking about whether or not we wanted to remove txs from the queue once we mock-stream them, but thought it might end up causing more problems if we have two stream threads concurrently

Reply:
thanks for catching this! let me clean up the comment :)